compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -149,44 +164,38 @@ internal class ProgressionLoopHeader(
                 //       // Loop body
                 //     } while (loopVar != last)
                 //   }
-                assert(loopVariable != null)
-                val booleanNotFun = context.irBuiltIns.booleanClass.functions.first { it.owner.name.asString() == "not" }
-                val newCondition = irCallOp(booleanNotFun, booleanNotFun.owner.returnType, irCall(context.irBuiltIns.eqeqSymbol).apply {
-                    putValueArgument(0, irGet(loopVariable!!))
-                    putValueArgument(1, irGet(last))
-                })
                 val newLoop = IrDoWhileLoopImpl(oldLoop.startOffset, oldLoop.endOffset, oldLoop.type, oldLoop.origin).apply {
                     label = oldLoop.label
-                    condition = newCondition
+                    condition = irNotEquals(irGet(loopVariable!!), irGet(last))
                     body = newBody
                 }
-                val notEmptyCheck = irIfThen(buildLoopCondition(builder), newLoop)
+                val notEmptyCheck = irIfThen(buildLoopCondition(this@with), newLoop)
                 LoopReplacement(newLoop, notEmptyCheck)
             } else {
-                // If the induction variable can NOT overflow, use a simple while loop. Loop is lowered into something like:
+                // If the induction variable can NOT overflow, use a do-while loop. Loop is lowered into something like:
                 //
-                //   while (inductionVar <= last) {
-                //       val loopVar = inductionVar
-                //       inductionVar += step
-                //       // Loop body
+                //   if (inductionVar <= last) {
+                //     do {
+                //         val loopVar = inductionVar
+                //         inductionVar += step
+                //         // Loop body
+                //     } while (inductionVar <= last)
                 //   }
-                val newLoop = IrWhileLoopImpl(oldLoop.startOffset, oldLoop.endOffset, oldLoop.type, oldLoop.origin).apply {
+                //
+                // Even though this can be simplified into a simpler while loop, using if + do-while (i.e., doing a loop inversion)
+                // performs better in benchmarks. In cases where `last` is a constant, the `if` may be optimized away.
```

Comment:
This is a mystery.

Do you have any idea why this is faster? It shouldn't be because of branching - the loop inversion in question would be performed anyway (without code duplication) by, e.g., MachineBlockPlacement in llvm. Is it because we don't remove empty loops, but do remove empty conditionals? In either case, it seems like this is a performance problem stemming from somewhere else and should be fixed there.

This is especially strange since you only perform a loop inversion for progressions, not for array loops.

Reply:
Unfortunately I don't have a good idea of why this is so. In any case, one of my goals is to converge on a single shared `ForLoopsLowering`, and for this I have to track the current `ForLoopsLowering` in K/N as closely as possible, which does have do the inversion at this stage. And K/N also does the inversion for progressions but not array loops. ¯\_(ツ)_/¯ My top priority is to ensure that when we switch over K/N to the shared version, that we are very confident that there aren't any performance regressions.

I am inclined to converge on the lowering first, and investigate this mystery later. Perhaps there is also a shortcoming in the existing benchmarks.